### PR TITLE
[DEV-169] Log / Block IP on sign in/up

### PR DESCRIPTION
**Pull request by Kolibri1990**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

File: src/shared/auth/ip.guard.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { CanActivate, ExecutionContext, ForbiddenException, Injectable } from '@nestjs/common';
+import { Config } from 'src/config/config';
+import { GeoLocationService } from 'src/integration/geolocation/geo-location.service';
+import { IpRepository } from 'src/subdomains/generic/user/models/ip/ip.repository';
+import { CountryService } from '../models/country/country.service';
+import * as requestIp from '@supercharge/request-ip';
```
Comment (davidleomay) [inline src/shared/auth/ip.guard.ts:6]: `import { getClientIp } from '@supercharge/request-ip';`

File: src/shared/auth/ip.guard.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+@Injectable()
+export class IpGuard implements CanActivate {
+  constructor(
+    private geoLocationService: GeoLocationService,
+    private countryService: CountryService,
+    private ipRepo: IpRepository,
+  ) {}
+  async canActivate(context: ExecutionContext): Promise<boolean> {
+    const req = context.switchToHttp().getRequest();
+    const ip = await requestIp.getClientIp(req);
```
Comment (davidleomay) [inline src/shared/auth/ip.guard.ts:17]: Unnecessary `await`

File: src/shared/auth/ip.guard.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    const ipObject = this.ipRepo.create({
+      ip,
+      country,
+      result,
+      url: req.url,
+      address,
+    });
+    await this.ipRepo.save(ipObject);
+    if (!result) throw new ForbiddenException('The country of IP address is not allowed');
+    return result;
```
Comment (davidleomay) [inline src/shared/auth/ip.guard.ts:30]: Throwing the error is not required, returning `false` will throw a ForbiddenException

File: src/shared/auth/ip.guard.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  private async checkIpCountry(userIp: string): Promise<{ country: string; result: boolean }> {
+    if (Config.environment === 'loc' || userIp?.includes(Config.azureIpSubstring))
+      return { country: 'INTERN', result: true };
+    const country = await this.geoLocationService.getCountry(userIp);
+    if (!country) throw new ForbiddenException('The country of IP address is unknown');
+    const countryObject = await this.countryService.getCountryWithSymbol(country);
+
+    return { country, result: countryObject?.ipEnable };
```
Comment (davidleomay) [inline src/shared/auth/ip.guard.ts:40]: Maybe just return country `undefined` and result `false`, if no country? Country should then be nullable in entity.

File: src/subdomains/generic/user/models/ip/ip.entity.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { IEntity } from 'src/shared/models/entity';
+import { Entity, Column } from 'typeorm';
+
+@Entity()
+export class Ip extends IEntity {
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/ip/ip.entity.ts:5]: Maybe rename it to SignInLog or similar?

**Review by davidleomay — COMMENTED** [review]
Copy changes to LOCK

File: src/shared/auth/ip.guard.ts (hunk 3 above)
Comment (Kolibri1990) [reply src/shared/auth/ip.guard.ts:30]: Specific error message

File: src/subdomains/generic/user/models/ip/ip.entity.ts (hunk 5 above)
Comment (Kolibri1990) [reply src/subdomains/generic/user/models/ip/ip.entity.ts:5]: renamed it to IpLog

File: migration/1677015014694-add-ip-logger.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" ADD CONSTRAINT "UQ_8a21f1713dbe0211d8493128774" UNIQUE ("signature")`);
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" ADD "mail" nvarchar(256)`);
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" ALTER COLUMN "address" nvarchar(256) NOT NULL`);
+    }
+
+    async down(queryRunner) {
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" ALTER COLUMN "address" nvarchar(256)`);
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" DROP COLUMN "mail"`);
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" DROP CONSTRAINT "UQ_8a21f1713dbe0211d8493128774"`);
+        await queryRunner.query(`ALTER TABLE "dbo"."wallet" DROP COLUMN "signature"`);
```
Comment (davidleomay) [inline migration/1677015014694-add-ip-logger.js:18]: Unwanted changes?

File: src/subdomains/generic/user/models/ip-log/ip-log.repository.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { EntityRepository, Repository } from 'typeorm';
+import { IpLog } from './ip-log.entity';
+
+@EntityRepository(IpLog)
+export class IpRepository extends Repository<IpLog> {}
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/ip-log/ip-log.repository.ts:5]: IpLogRepository

**Review by davidleomay — APPROVED** [review]

**Pull request merged**